R/rtf_rich_text.R

```diff
@@ -0,0 +1,116 @@
+#    Copyright (c) 2023 Merck & Co., Inc., Rahway, NJ, USA and its affiliates. All rights reserved.
+#
+#    This file is part of the r2rtf program.
+#
+#    r2rtf is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU General Public License as published by
+#    the Free Software Foundation, either version 3 of the License, or
+#    (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#' Text to formatted RTF Encode
+#'
+#' @param text Plain text.
+#' @param theme Named list defining themes for tags. See \code(rtf_text) for
+#' details on possible formatting.
+#'
+#' @section Specification:
+#' \if{latex}{
+#'  \itemize{
+#'    \item Validate if theme list items correspond to \code{font_type()} arguments.
+#'    \item Create regex expressions to match `{}` and `.tag` in text.
+#'    \item Extract tagged text from input text.
+#'    \item Extract tags from tagged text.
+#'    \item Extract text from tagged text.
+#'    \item Validate that lengths of extractions are all the same.
+#'    \item Validate that tags are defined in the `theme` argument.
+#'    \item Execute `rtf_text` with extracted text and relevant formatting.
+#'    \item Reinsert encoded formatted text to original input text.
+#'  }
+#'  }
+#' \if{html}{The contents of this section are shown in PDF user manual only.}
+#'
+#' @importFrom checkmate assert_choice assert_true
+#'
+#' @export
+#'
+#' @examples
+#' rtf_rich_text()
+rtf_rich_text <- function(text = "This is {.emph important}. This is {.blah relevant}.",
+                          theme = list(
```

Comment:
Could we extend the theme to commonly used format? 

not sure `.emph` should always be `blue` and `.blah` should be `red`.

Reply:
Can you be more specific here? What defaults would you like to suggest?

Reply:
Feel a combination of the style in the link below + color name allowed by RTF should be sufficient 

## text format: 
https://merck.github.io/r2rtf/articles/rtf-text.html#text-format

## color: 
https://www.rdocumentation.org/packages/simplecolors/versions/0.1.1/topics/color_table
https://merck.github.io/r2rtf/articles/rtf-text.html#text-color

## example

if we want a text to be bold and blue, we should have 

"this is {.emph .blue important} that is {.blah .red relevant}"

Reply:
Can you give further arguments as to why we should enable multiple tags per field when a single tag can be defined to cover everything that the user needs. Otherwise we'll need to also do checks to make sure that the tags theme definitions don't overlap (two tags that both declare a color argument...)

Reply:
Just a small random fact: `emph` is the [semantic](https://cli.r-lib.org/reference/inline-markup.html) way of saying italic text, not bold text (in most contexts).

I agree that the styles should be composable - but not pre-composed. One style + [custom theme](https://cli.r-lib.org/reference/themes.html) (cli's approach) will surely work. The stacked styles also looks interesting enough - you can easily check the viability by seeing if glue supports that sort of syntax.

In cli's theme approach, you can have only one value per property (for example, color) defined, because the list properties (component) name cannot have duplicates.

Reply:
I see the point. The single keyword is a better design. Thanks.

Reply:
So I understand that "one style" is corresponding to ["a single cli inline class"](https://cli.r-lib.org/reference/inline-markup.html#classes).
And then the idea is that "a custom theme" would correspond to how cli updates their classes 
```
# .emph already exists somewhere as it's a default inline class
# the following code just updates that definition to include the color red
cli_div(theme = list(span.emph = list(color = "red")))
adjective <- "nice and red"
cli_text("This is {.emph {adjective}}.")
```

This would recommend a structure similar to what I've made, but:
- default classes (`emph`, `strong`, ... any else?) which are defined internally and are only updated (additively) when the user provides a `theme` argument.
- a matching/updating step that allows the user's input of `theme = list(.emph = list(color = "red"))` to update the already existing `.emph` class
- still limit to a single class per `{}`

Reply:
Feel `emph` and `strong` are sufficient. 

As long as the structure is there, we can add additional format as needed.

Let's also add an vignettes to help user understand how to customize format. e.g. what if a user want a bold purple text with underline.